core/src/main/scala/kafka/log/Log.scala

```diff
@@ -551,6 +556,22 @@ class Log(@volatile var dir: File,
     }
   }
 
+  def renameDir(name: String): Boolean = {
+    lock synchronized {
+      val renamedDir = new File(dir.getParent, name)
+      val renameSuccessful = renamedDir == dir || dir.renameTo(renamedDir)
```

Comment:
We should probably use `Utils.atomicMoveWithFallback`.

Reply:
I have considered this API. `Utils.atomicMoveWithFallback` calls `Files.move`. And according to the Java doc of `Files.move`, this method will fail if the target file exists, which is the case here when we replace the current replica with the future replica.

Reply:
`Files.move` does not fail if the target file already exists given the right `CopyOptions`, which `atomicMoveWithFallback` uses.

Reply:
In particular, `ATOMIC_MOVE` should work on Linux. And on Windows, we fallback to `REPLACE_EXISTING`.

Reply:
Although, this is a directory, so not sure. Maybe we should write a test to verify. Either way, if `ATOMIC_MOVE` doesn't work, we should use `Files.move` with `REPLACE_EXISTING`, which should work, I believe.

Reply:
@ijuma I made a mistake previously. The `renameDir` will rename a directory but the destination direction does not exist. Thanks for the explanation. I will try and test this approach.

Reply:
Sounds good. The main advantage of the `Files` APIs when compared to the legacy ones is that they provide error messages when things go wrong. That can be quite helpful.

Reply:
I have updated the patch as suggested.